activejob/lib/active_job/test_helper.rb

```diff
@@ -402,6 +402,18 @@ def assert_enqueued_with(job: nil, args: nil, at: nil, queue: nil, &block)
       end
 
       matching_job = jobs.find do |enqueued_job|
+        if expected_args[:job].present?
+          if expected_args[:job].respond_to?(:call)
+            if !expected_args[:job].call(enqueued_job[:job])
+              potential_matches << enqueued_job
+              next
+            end
+          elsif expected_args[:job] != enqueued_job[:job]
+            potential_matches << enqueued_job
+            next
+          end
+        end
```

Comment:
It seems like this doesn't solve the problem in the general case, since an `ActiveJob::DeserializationError` can still blow up the assertion if `:job` isn't specified, no?

Also, I think `assert_performed_with` needs similar protection.  Something like the following would cause the same problem:

```ruby
model = Model.create!
ModelJob.perform_later(model)
perform_enqueued_jobs
model.destroy
assert_performed_with(job: ModelJob)
```

Perhaps we could rescue `ActiveJob::DeserializationError` in [`deserialize_args_for_assertion`](https://github.com/rails/rails/blob/1a7be948426a7300a8ea5a2cdb4c55d645b2552c/activejob/lib/active_job/test_helper.rb#L694)?  Although I'm not clear on whether we should attempt to [match](https://github.com/rails/rails/blob/1a7be948426a7300a8ea5a2cdb4c55d645b2552c/activejob/lib/active_job/test_helper.rb#L408-L414) such jobs afterwards or not... :thinking:

Reply:
Yeah - this definitely isn't a "fix all scenarios" patch, just make it less likely to fail 😓 

The new scenario you raise is a super interesting one (and solving that might make this a more robust solution). I'll pick this back up later in the week and see what I can come up with.

Thanks for taking a look and giving some feedback, @jonathanhefner!